specs/newton/approved/initial_design.rst

```diff
@@ -0,0 +1,31 @@
+..
+ This work is licensed under a Creative Commons Attribution 3.0 Unported
+ License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+
+========================
+Initial design of Ephyra
+========================
+
+Ephyra is a simple ...
+
+Problem description
+===================
+
+Imagine that you have a pool of Ironic nodes running a Hadoop workload. You would like to do Continuous Deployment of the control plane for this OpenStack deployment, while having a graceful rollback procedure for when the new version of the control plane is broken for some reason. Specifically for this thought experiment, you can assume that only Ironic is supported as a Nova hypervisor in this deployment, and that the workload running on those Ironic nodes can handle partial outages without degrading performance.
+
+The solution we have come to is that you have two OpenStack deployments. In the initial state, the first of these deplouments (which we shall call CP1) has all of the Ironic nodes registered with it, and the second deployment (CP2) has none.
+
+The Continous Deployment system detects that there is a new version of OpenStack to deploy. It installs that version onto CP2 in whatever manner it desires. Installation of OpenStack is outside the scope of Ephyra, but it is assumed to be automated.
+
+An API call is then made to Ephyra, which manages migrating Ironic nodes from CP1 to CP2. This is done by marking each node in turn as in maintenance mode in Hadoop, shutting down the instance on the node, de-enrolling the node from Ironic in CP1, enrolling the node with Ironic in CP2, booting an instance mapping to that nodes "role" in Hadoop, and then informing Hadoop that the node is no longer in maintenance mode. This process is discussed in more detail below.
```

Comment:
Firstly, I think you should call them "blue" and "green" rather than CP1 and CP2.

Secondly, we should make it clear that we should be pushing one type of change at a time - either a new golden image, or a new control plane, not both at the same time.

Thirdly, we need to document that Ephyra needs to provide an API call to indicate whether blue or green is the current active control plane.  This will be consumed by the CI/CD system only.